src/network/network.cpp

```diff
@@ -7,201 +7,219 @@
-    if (node.ip.empty()) {
-      LOG(log_wr_) << "Boot node ip is empty:" << node.ip << ":" << node.tcp_port << ":" << node.udp_port;
-      continue;
-    }
-    if (node.tcp_port == 0 || node.tcp_port > 65535) {
-      LOG(log_wr_) << "Boot node port invalid: " << node.tcp_port;
-      continue;
-    }
-    if (node.udp_port == 0 || node.udp_port > 65535) {
-      LOG(log_wr_) << "Boot node port invalid: " << node.udp_port;
+  LOG(log_nf_) << "Read Network Config: " << std::endl << conf_ << std::endl;
+
+  // TODO make all these properties configurable
+  dev::p2p::NetworkConfig net_conf;
+  net_conf.listenIPAddress = conf_.network_address;
+  net_conf.listenPort = conf_.network_tcp_port;
+  net_conf.discovery = true;
+  net_conf.allowLocalDiscovery = true;
+  net_conf.traverseNAT = false;
+  net_conf.publicIPAddress = {};
+  net_conf.pin = false;
+
+  TaraxaNetworkConfig taraxa_net_conf;
+  taraxa_net_conf.ideal_peer_count = conf_.network_ideal_peer_count;
+  // TODO conf_.network_max_peer_count -> conf_.peer_count_stretch
+  taraxa_net_conf.peer_stretch = conf_.network_max_peer_count / conf_.network_ideal_peer_count;
+  taraxa_net_conf.is_boot_node = conf_.network_is_boot_node;
+  for (auto const &node : conf_.network_boot_nodes) {
+    Public pub(node.id);
+    if (pub == key.pub()) {
+      LOG(log_wr_) << "not adding self to the boot node list";
       continue;
     }
+    LOG(log_nf_) << "Adding boot node:" << node.ip << ":" << node.tcp_port;
     auto ip = resolveHost(node.ip, node.tcp_port);
-    host_->addBootNode(dev::Public(node.id),
-                       dev::p2p::NodeIPEndpoint(ip.second.address(), node.udp_port, node.tcp_port));
-    boot_node_added++;
+    boot_nodes_[pub] = dev::p2p::NodeIPEndpoint(ip.second.address(), node.tcp_port, node.tcp_port);
   }
-  LOG(log_nf_) << " Number of boot node added: " << boot_node_added << std::endl;
-  host_->start(boot_node);
-  LOG(log_nf_) << "Started Network address: " << conf_.network_address << ":" << conf_.network_tcp_port << " :"
-               << conf_.network_udp_port << std::endl;
-  LOG(log_nf_) << "Started Node id: " << host_->id();
+  LOG(log_nf_) << " Number of boot node added: " << boot_nodes_.size() << std::endl;
+  string net_version = "TaraxaNode";  // TODO maybe give a proper name?
+  dev::bytes networkData;
+  if (!network_file_.empty()) {
+    networkData = contents(network_file_.string());
+  }
+  if (!networkData.empty()) {
+    host_ = std::make_shared<dev::p2p::Host>(net_version, net_conf, move(taraxa_net_conf), &networkData);
+  } else {
+    host_ = std::make_shared<dev::p2p::Host>(net_version, makeENR(key, net_conf), net_conf, move(taraxa_net_conf));
+  }
+  taraxa_capability_ = std::make_shared<TaraxaCapability>(*host_, tp_.unsafe_get_io_context(), conf_, db, pbft_mgr,
+                                                          pbft_chain, vote_mgr, next_votes_mgr, dag_mgr, dag_blk_mgr,
```

Comment:
just use reference to tp_ inside TaraxaCapability instead f io_context. Much cleaner and easier to understand